### ISSUE TITLE
Rule enforcing typeAnnotation on a complex query is correct

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint-community/eslint-plugin-eslint-plugin#340]
Copying conversation from https://github.com/typescript-eslint/typescript-eslint/issues/4065#issuecomment-955397808 & https://github.com/typescript-eslint/typescript-eslint/pull/6444#discussion_r1102035077: typescript-eslint doesn't have a way to infer the types of nodes when the ESQuery selector is complex:

```ts
return {
  'ClassDeclaration, ClassExpression[id]'(
    node
    // ^? TSESTree.Node
    // Expected: something like...
    // TSESTree.ClassDeclaration | TSESTree.ClassExpression & { id: TSESTree.Identifier }
  ) {
```

Even with these two not(-yet?)-implemented helpers (neither of which may ever be useful for public consumers):
* https://github.com/typescript-eslint/typescript-eslint/issues/4065
* https://github.com/typescript-eslint/typescript-eslint/issues/6224

...there's no way for custom rules written in TypeScript to enforce that the type of `node` matches its AST shape. Which means devs can write blatantly wrong code like:

```ts
return {
  'ClassDeclaration, ClassExpression[id]'(
    node: TSESTree.Identifier
```

Is `eslint-plugin-eslint-plugin` an appropriate place to write a lint rule that makes sure the type annotation on a node matches the node itself?

This rule might be tricky to write given that https://github.com/microsoft/TypeScript/issues/9879 blocks having a TypeScript API to check type assignability. The rule could use basic string comparisons in the meantime.

**Comment by bmish** [comment]
I'm open to accepting rules targeted at TypeScript rules/syntax, including this one.

I previously added general [compatibility](https://github.com/eslint-community/eslint-plugin-eslint-plugin/pull/197) with TypeScript rules in [v4](https://github.com/eslint-community/eslint-plugin-eslint-plugin/releases/tag/v4.0.0).

Any new rule targeted at TypeScript rules should just no-op when running against a JavaScript rule.

We can even use `meta.docs.category: "TypeScript"` to have a separate group/list of our TypeScript-focused rules (related: #341).
